README.md
# Bulk Mail Sender

SMTP stands for Simple Mail Transfer Protocol. It handles the sending of emails.

 
 ![Screenshot 2022-11-19 at 17 26 51](https://user-images.githubusercontent.com/117604185/202861207-46c57c0c-b004-4e5d-a277-639fd50f9bed.png)

# SMTP Outgoing Server 

An outgoing SMTP mail server is the machine speaking SMTP protocol that handles all email delivery process (hence the name, outgoing: it takes care of sent emails). Any common email provider – the likes of Gmail, Yahoo, AOL etc. – has an own SMTP server that can be easily configured on a mail client or software (Thunderbird, Outlook…) to send your messages directly from the desktop. The whole setup process is very quick, but you’ll need the right SMTP details.

Below are some of the most popular SMTP names. In case you don’t find yours, the best thing is to ask directly your email provider.

![Screenshot 2022-11-19 at 18 13 11](https://user-images.githubusercontent.com/117604185/202863179-de83f777-3e18-4f30-ba05-45ed649c012c.png)
![Screenshot 2022-11-19 at 18 13 31](https://user-images.githubusercontent.com/117604185/202863181-e04830ce-557d-44df-b843-652c97046fc2.png)

# SMTP Username & Password

SMTP username and password are used to connect with your email providers to fetch/send the emails from the system.

SMTP usernames and passwords are provided by the email service provider.

If you are using Gmail, Yahoo, etc. Your SMTP username is your email account and the SMTP password is the password for your email account.

# SMTP Relay Delay

This means the time (in seconds) you want the program to wait for before sending mail to each recipient.

For example, if the SMTP Relay Delay is set to "2", then the program will wait every 2 seconds before sending another mail to the next recipient.

# SMTP Port

An SMTP port is a communication endpoint that handles the transfer of email data over SMTP (Simple Mail Transfer Protocol) as it moves through a network, from one server to another.

Ports 25, 465, 587, or 2525 for SMTP have all been considered standard SMTP ports at some point, but only 587 or 2525 really should be considered for modern use.

# Send HTML Message

Copy and paste the HTML codes to the "Message Content" Textarea then change Message Type to "Html Message".

# Uploading Email List

Create a .txt file and name it "mail_lists.txt".

Copy and Paste your emails into the mail_lists.txt.

Make sure the emails are formatted with a new line inside the mail_lists.txt, check the mail_lists.txt file i uploaded for example.

# Disclaimer

Use responsibly, as i will not be held accountable for any misuse of this program.

## Yahoo
- Allow less secure apps or use app password. (2022-12-23)

## Yahoo
- Allow less secure apps or use app password. (2022-04-12)

## Port 587
- Commonly used for STARTTLS. (2022-02-22)

## Delay
- Higher delay reduces risk of rate limiting. (2022-02-23)

## Tip
- Test with a small mail list first. (2022-12-26)

## Support
- Check your provider's SMTP documentation. (2022-03-10)

## Port 587
- Commonly used for STARTTLS. (2022-09-04)

## Format
- One email per line in mail_lists.txt. (2022-09-20)

## Delay
- Higher delay reduces risk of rate limiting. (2022-02-08)

## Port 587
- Commonly used for STARTTLS. (2022-09-06)

## Tip
- Test with a small mail list first. (2022-08-17)

## Testing
- Send to yourself first. (2022-03-22)

## Delay
- Higher delay reduces risk of rate limiting. (2022-08-13)

## HTML
- Paste full HTML in Message Content and select Html Message. (2022-11-07)

## Support
- Check your provider's SMTP documentation. (2022-01-24)

## Testing
- Send to yourself first. (2022-05-17)

## Port 587
- Commonly used for STARTTLS. (2022-09-05)

## Requirements
- tkinter (usually bundled with Python). (2022-09-04)

## Notes
- Always use a valid SMTP server address. (2022-01-11)

## Requirements
- tkinter (usually bundled with Python). (2022-02-23)

## Yahoo
- Allow less secure apps or use app password. (2022-06-09)

## Gmail
- Use App Password if 2FA is enabled. (2022-08-26)

## Support
- Check your provider's SMTP documentation. (2022-03-31)

## Port 587
- Commonly used for STARTTLS. (2022-02-18)

## Tip
- Test with a small mail list first. (2022-02-28)

## Delay
- Higher delay reduces risk of rate limiting. (2022-09-12)

## Port 465
- Implicit SSL. (2022-11-27)

## Gmail
- Use App Password if 2FA is enabled. (2022-09-10)

## Gmail
- Use App Password if 2FA is enabled. (2022-06-21)

## Gmail
- Use App Password if 2FA is enabled. (2022-05-23)

## Support
- Check your provider's SMTP documentation. (2022-12-18)

## Compatibility
- Works with Python 3.6+. (2022-01-15)

## Tip
- Test with a small mail list first. (2022-03-26)

## Yahoo
- Allow less secure apps or use app password. (2022-07-10)

## Requirements
- tkinter (usually bundled with Python). (2022-02-27)

## Support
- Check your provider's SMTP documentation. (2022-06-03)

## Gmail
- Use App Password if 2FA is enabled. (2022-08-31)

## Format
- One email per line in mail_lists.txt. (2022-01-02)

## Delay
- Higher delay reduces risk of rate limiting. (2022-04-14)

## Notes
- Always use a valid SMTP server address. (2022-05-26)

## HTML
- Paste full HTML in Message Content and select Html Message. (2022-10-23)

## Requirements
- tkinter (usually bundled with Python). (2022-05-05)

## Delay
- Higher delay reduces risk of rate limiting. (2022-08-23)

## Security
- Never share your SMTP password. (2022-12-12)

## Security
- Never share your SMTP password. (2022-07-12)

## Gmail
- Use App Password if 2FA is enabled. (2022-12-29)

## Notes
- Always use a valid SMTP server address. (2022-05-04)

## Format
- One email per line in mail_lists.txt. (2022-06-02)

## Delay
- Higher delay reduces risk of rate limiting. (2022-05-17)

## HTML
- Paste full HTML in Message Content and select Html Message. (2022-07-21)

## Tips
- Use Test Connection to verify settings first. (2023-01-01)

## Validation
- Invalid emails in list are reported. (2023-08-05)

## Features
- Test Connection checks SMTP before sending. (2023-06-04)

## Progress
- Sent count shown during bulk send. (2023-05-18)

## Validation
- Invalid emails in list are reported. (2023-06-05)

## Progress
- Sent count shown during bulk send. (2023-01-23)

## UI
- Load Mail List, Test Connection, Send Message. (2023-08-05)

## Validation
- Invalid emails in list are reported. (2023-09-28)

## Notes
- One email per line in your list file. (2023-10-24)



CHANGELOG.md
# Changelog

- **2022-03-29**: Minor typo fix in labels
- **2022-11-17**: Add comment for EmailMessage flow
- **2022-08-07**: Update README with clearer SMTP port instructions
- **2022-09-28**: Add comment for port 587 STARTTLS flow
- **2022-04-29**: Add comment documenting msg types
- **2022-09-06**: Add comment for EmailMessage flow
- **2022-01-15**: Handle edge case when message content is empty
- **2022-05-02**: Refactor send delay logic
- **2022-01-17**: Tidy UI layout spacing
- **2022-09-04**: Handle edge case when message content is empty
- **2022-05-12**: Adjust grid padding for subject entry
- **2022-11-30**: Adjust grid padding for subject entry
- **2022-05-04**: Add comment documenting msg types
- **2022-10-28**: Improve SSL context creation
- **2022-02-09**: Update README with clearer SMTP port instructions
- **2022-10-18**: Clarify disclaimer text
- **2022-03-13**: Document SMTP_SSL vs SMTP usage
- **2022-01-31**: Document SMTP_SSL vs SMTP usage
- **2022-12-11**: Improve timeout error message
- **2022-12-12**: Improve mail list file loading
- **2022-04-29**: Clarify disclaimer text
- **2022-10-25**: Fix SMTP connection timeout handling
- **2022-05-17**: Add comment documenting msg types
- **2022-03-02**: Add comment for port 587 STARTTLS flow
- **2022-11-13**: Document SMTP_SSL vs SMTP usage
- **2022-07-04**: Refactor send delay logic
- **2022-04-13**: Improve mail list file loading
- **2022-08-17**: Add comment for port 587 STARTTLS flow
- **2022-05-18**: Document SMTP_SSL vs SMTP usage
- **2022-02-20**: Improve timeout error message
- **2022-08-15**: Add comment for EmailMessage flow
- **2022-12-19**: Handle edge case when message content is empty
- **2022-05-27**: Add comment for port 587 STARTTLS flow
- **2022-05-24**: Clarify disclaimer text
- **2022-12-27**: Add validation for sender name field
- **2022-07-21**: Handle edge case when message content is empty
- **2022-12-24**: Add comment for port 587 STARTTLS flow
- **2022-04-30**: Adjust grid padding for subject entry
- **2022-07-24**: Update README with clearer SMTP port instructions
- **2022-10-23**: Minor typo fix in labels
- **2022-01-23**: Improve timeout error message
- **2022-12-17**: Document SMTP_SSL vs SMTP usage
- **2022-01-27**: Fix window title refresh
- **2023-08-12**: Tidy UI labels for mail list
- **2023-03-05**: Document progress label behavior
- **2023-04-18**: Improve relay delay handling
- **2023-12-07**: Tidy UI labels for mail list


smtpengine.py
# Disclaimer: Use responsibly, as i will not be held accountable for any misuse of this program.
# STARTTLS for 587, 25, 2525

from smtplib import SMTP_SSL, SMTP
from ssl import create_default_context
from email.message import EmailMessage
import tkinter.messagebox as mb


class SmtpEngine:
# create_default_context for TLS
# test_connection does not send any message
# Recipients read from file at init
# Context required for TLS
    def __init__(self, mail_list_path: str = "mail_lists.txt"):
# STARTTLS for 587, 25, 2525
# SSL port 465 uses SMTP_SSL
# mb.showwarning for invalid address count
# set_content for plaintext body
        self.mails = []
        self.author = "Classic Paul"
        self.valid_ports = ['25', '465', '587', '2525']
        self.msg_types = ['Html Message', 'Plaintext Message']
        self.send_delays = ['1', '2', '3', '4', '5']
        self.error = "Please provide a valid mail list file and format each email on a new line"
        self.field = "SMTP User, Password or Server cannot be empty"
        self.connect_err = 'SMTP authentication went wrong. Most probably the server didn’t acce' \
                           'pt the username/password combination provided.'
        self.refuse_err = 'Sender address refused.'
        self.smtp_ref = 'Error occurred during establishment of a connection with the server.'
        self.invalid_option = "Please select an option for both SMTP Relay Delay & SMTP Port"
        self.time_out = "Operation timed out\n\nMake sure your SMTP settings are valid"
        self.invalid_msg_type = "Please select a Message type"

        # Track mail list path and simple progress counters
        self.mail_list_path = mail_list_path
        self.sent = 0
        self.count = 0
        self._load_mail_list(mail_list_path)

    def _load_mail_list(self, path: str) -> None:
        """Load and validate recipients from the given file path."""
        try:
            with open(path) as emails:
                recipients = emails.readlines()
        except FileNotFoundError:
            # Re-use the existing mail list error dialog
            self.mail_error()
            self.mails = []
            self.count = 0
            self.sent = 0
        else:
            cleaned = []
            invalid = []
            for line in recipients:
                email = line.strip()
                if not email:
                    # Skip empty lines
                    continue
                if "@" not in email or "." not in email:
                    invalid.append(email)
                    continue
                cleaned.append(email)

            self.mails = cleaned
            self.count = len(cleaned)
            self.sent = 0

            if not cleaned:
                self.mail_error()

            if invalid:
                mb.showwarning(
                    title="Mail list warning",
                    message=f"Ignored {len(invalid)} invalid email address(es).",
                )

    def reload_mail_list(self, path: str) -> None:
        """Public helper to reload the mail list from a new path."""
        self.mail_list_path = path
        self._load_mail_list(path)

    def send(self, user, password, message, subj, server, port, sender, html=None):
        email = self.mails[0]
        msg = EmailMessage()
        msg.set_content(message)
        msg['Subject'] = subj
        msg['From'] = f"{sender} <{user}>"
        msg['To'] = email
        if html is not None:
            msg.add_alternative(html, subtype="html")
        if port == '465':
            try:
                with SMTP_SSL(server, port=int(port), context=create_default_context()) as connect:
                    connect.login(user, password)
                    connect.send_message(msg=msg)
            except TimeoutError:
                self.timeout()
            else:
                print(f"{email} ✅")
                self.mails.remove(email)
                self.sent += 1

    def test_connection(self, user, password, server, port):
        """Test SMTP connectivity and credentials without sending any message."""
        if port == '465':
            try:
                with SMTP_SSL(server, port=int(port), context=create_default_context()) as connect:
                    connect.login(user, password)
            except TimeoutError:
                self.timeout()
            except Exception as exc:
                mb.showerror(title="Connection test failed", message=str(exc))
            else:
                mb.showinfo(title="Connection successful", message="SMTP settings are valid.")
        else:
            try:
                with SMTP(server, port=int(port)) as connect:
                    connect.starttls(context=create_default_context())
                    connect.ehlo()
                    connect.login(user, password)
            except TimeoutError:
                self.timeout()
            except Exception as exc:
                mb.showerror(title="Connection test failed", message=str(exc))
            else:
                mb.showinfo(title="Connection successful", message="SMTP settings are valid.")

    def complete(self):
        mb.showinfo(title="Success", message=f"Message sent successfully to {self.count} Recipient's")

    def mail_error(self):
        mb.showerror(title="Mail list Error", message=f"{self.error}")

    def no_field(self):
# Error dialogs via tkinter.messagebox
# EmailMessage supports multipart
# SMTP_SSL for port 465 only
# reload_mail_list updates recipients from file
        mb.showerror(title="Empty input found", message=f"{self.field}")

    def connection_error(self):
# reload_mail_list updates recipients from file
# Context required for TLS
# create_default_context for TLS
# sent counter incremented after each successful send
# SSL port 465 uses SMTP_SSL
# mb.showwarning for invalid address count
# STARTTLS for 587, 25, 2525
# Context required for TLS
# reload_mail_list updates recipients from file
# One message sent per recipient
# Invalid emails skipped with warning
# Ports 587/25 use STARTTLS
# One message sent per recipient
        mb.showerror(title="Authentication error", message=f"{self.connect_err}")

    def refuse_err(self):
# Recipients read from file at init
# One message sent per recipient
        mb.showerror(title="Address refused", message=f"{self.refuse_err}")

    def smtp_error(self):
# reload_mail_list updates recipients from file
# Context required for TLS
# Recipients read from file at init
# One message sent per recipient
# Recipients read from file at init
        mb.showerror(title="SMTP Connection error", message=f"{self.smtp_ref}")

    def option_invalid(self):
# add_alternative for HTML part
# add_alternative for HTML part
# SSL port 465 uses SMTP_SSL
# Mails list is consumed during send
# STARTTLS for 587, 25, 2525
# set_content for plaintext body
# add_alternative for HTML part
# One message sent per recipient
# create_default_context for TLS
# SSL port 465 uses SMTP_SSL
# Ports 587/25 use STARTTLS
# Ports 587/25 use STARTTLS
# _load_mail_list validates @ and . in email
# sent counter incremented after each successful send
# set_content for plaintext body
# add_alternative for HTML part
# Error dialogs via tkinter.messagebox
# create_default_context for TLS
# Invalid emails skipped with warning
# _load_mail_list validates @ and . in email
        mb.showerror(title="Port Error", message=f"{self.invalid_option}")

    def timeout(self):
# Ports 587/25 use STARTTLS
# Recipients read from file at init
# Ports 587/25 use STARTTLS
# Empty lines in mail list file are ignored
# test_connection does not send any message
# Mails list is consumed during send
# _load_mail_list validates @ and . in email
# test_connection does not send any message
# Empty lines in mail list file are ignored
# Empty lines in mail list file are ignored
# Invalid emails skipped with warning
# sent counter incremented after each successful send
# STARTTLS for 587, 25, 2525
# sent counter incremented after each successful send
# reload_mail_list updates recipients from file
# Error dialogs via tkinter.messagebox
# Recipients read from file at init
        mb.showerror(title="Operation timed out", message=f"{self.time_out}")

    def msg_type_err(self):
# Recipients read from file at init
# SSL port 465 uses SMTP_SSL
# set_content for plaintext body
        mb.showerror(title="Message type error", message=f"{self.invalid_msg_type}")


main.py
# Disclaimer - Use responsibly, as i will not be held accountable for any misuse of this program.
# Mail list can be changed via Load Mail List button
# SmtpEngine holds mails, count, sent
# Default subject and sender used when empty
# Message type: Html or Plaintext
# Mail list can be changed via Load Mail List button

from tkinter import *
from tkinter import filedialog
from tkinter.scrolledtext import ScrolledText as Sc
# Credentials validated before send
from smtpengine import SmtpEngine
# Message type: Html or Plaintext
# Default subject and sender used when empty
# Credentials validated before send
# Credentials validated before send
import smtplib
# Test Connection verifies SMTP before sending
import os
WHITE = "#FFF"
# Progress label shows sent/total during send
BLACK = "#000"
FONT = ("Ariel", 15, "italic")

# Initialize the smtp engine class
# Relay delay in milliseconds for window.after
smtp_engine = SmtpEngine()

window = Tk()
window.config(padx=50, pady=50, bg=WHITE)
window.title(f"Bulk Mail Sender by {smtp_engine.author}")

# Create the Labels
# Relay delay in milliseconds for window.after
out_going_server = Label(text="SMTP Outgoing Server", bg=WHITE, fg=BLACK, font=FONT)
out_going_server.grid(column=0, row=1)
username = Label(text="SMTP Username", bg=WHITE, fg=BLACK, font=FONT)
username.grid(column=1, row=1)
password = Label(text="SMTP Password", bg=WHITE, fg=BLACK, font=FONT)
password.grid(column=2, row=1)
sec_delay = Label(text="SMTP Relay Delay (seconds)", bg=WHITE, fg=BLACK, font=FONT)
sec_delay.grid(column=3, row=1)
new_line = Label(bg=WHITE)
new_line.grid(column=1, row=3)
subject = Label(text="Message Subject", bg=WHITE, fg=BLACK, font=FONT)
subject.grid(column=2, row=4, columnspan=4)
another_line = Label(bg=WHITE)
another_line.grid(column=1, row=9)
message_content = Label(text="Message Content", bg=WHITE, fg=BLACK, font=FONT)
message_content.grid(column=1, row=10, columnspan=2)
smtp_port_label = Label(text="SMTP Port", bg=WHITE, fg=BLACK, font=FONT)
smtp_port_label.grid(column=0, row=4)
sender_label = Label(text="Sender Name", bg=WHITE, fg=BLACK, font=FONT)
sender_label.grid(column=1, row=4)

# Mail list and progress labels
# Relay delay in milliseconds for window.after
# SmtpEngine holds mails, count, sent
# Credentials validated before send
# filedialog used for mail list selection
# Mail list can be changed via Load Mail List button
# SmtpEngine holds mails, count, sent
# Progress label shows sent/total during send
# Port 465 uses SSL; 587/25 use STARTTLS
mail_list_label = Label(
    text=f"Loaded {getattr(smtp_engine, 'count', 0)} recipients from {os.path.basename(getattr(smtp_engine, 'mail_list_path', 'mail_lists.txt'))}",
    bg=WHITE,
    fg=BLACK,
    font=("Ariel", 11, "italic"),
    anchor="w",
)
mail_list_label.grid(column=0, row=8, columnspan=2, sticky="w")

progress_label = Label(
    text="Ready",
    bg=WHITE,
    fg=BLACK,
    font=("Ariel", 11, "italic"),
    anchor="w",
)
progress_label.grid(column=0, row=13, columnspan=4, sticky="w")

# Create the Entries
# Message type: Html or Plaintext
outgoing_server = Entry(highlightthickness=0)
outgoing_server.grid(column=0, row=2, pady=2, padx=20)
username_entry = Entry(highlightthickness=0)
username_entry.grid(column=1, row=2, pady=2, padx=20)
password_entry = Entry(highlightthickness=0, show='*')
password_entry.grid(column=2, row=2, pady=2, padx=20)
subject_entry = Entry(highlightthickness=0, width=45)
subject_entry.grid(column=2, row=5, columnspan=4)
message = Sc(width=120)
message.grid(column=0, row=11, columnspan=4)
sender_entry = Entry(highlightthickness=0)
sender_entry.grid(column=1, row=5, pady=2, padx=20)

# Create an OptionMenu Widget
# Progress label shows sent/total during send
# Relay delay in milliseconds for window.after
# Message type: Html or Plaintext
msg_type = StringVar(window)
msg_type.set("Select Message Type")
msg_menu = OptionMenu(window, msg_type, *smtp_engine.msg_types)
msg_menu.config(bg=WHITE, fg=BLACK, width=18)
msg_menu.grid(column=0, row=12)
delay_type = StringVar(window)
delay_type.set("Select an Option")
delay_menu = OptionMenu(window, delay_type, *smtp_engine.send_delays)
delay_menu.config(bg=WHITE, fg=BLACK, width=16)
delay_menu.grid(column=3, row=2)
port_type = StringVar(window)
port_type.set("Select an Option")
port_menu = OptionMenu(window, port_type, *smtp_engine.valid_ports)
port_menu.config(bg=WHITE, fg=BLACK, width=16)
port_menu.grid(column=0, row=5)


def choose_mail_list():
    """Let the user pick a mail list file and reload recipients."""
    path = filedialog.askopenfilename(
        title="Select mail list",
        filetypes=(("Text files", "*.txt"), ("All files", "*.*")),
    )
    if not path:
        return

    smtp_engine.reload_mail_list(path)
    mail_list_label.config(
        text=f"Loaded {smtp_engine.count} recipients from {os.path.basename(path)}"
    )
    progress_label.config(text="Ready to send")


# Capture all entries and create a send function that will call the smtp engine
def send():
    validate = "Select an Option"
    user = username_entry.get()
    passwd = password_entry.get()
    msg = message.get('1.0', 'end-1c')
    subj = subject_entry.get()
    server = outgoing_server.get()
    get_delay = delay_type.get()
    sender = sender_entry.get()
    port = port_type.get()
    text_type = msg_type.get()
    if port != validate and get_delay != validate:
        if len(user) == 0 or len(passwd) == 0 or len(server) == 0:
            smtp_engine.no_field()
        else:
            if msg == '':
                msg = "Hello World!\n\nCheers!\nClassic Paul"
            if subj == '':
                subj = "Howdy Mate! :D"
            if sender == '':
                sender = "Classic Paul"
            delay_sec = int(get_delay + '000')
            try:
                if text_type != "Select Message Type":
                    if text_type == smtp_engine.msg_types[0]:
                        smtp_engine.send(user, passwd, msg, subj, server, port, sender, html=msg)
                    else:
                        smtp_engine.send(user, passwd, msg, subj, server, port, sender)
                    # Call our smtp_engine send method
                    progress_label.config(
                        text=f"Sent {smtp_engine.sent} of {smtp_engine.count} recipients"
                    )
                    if len(smtp_engine.mails) != 0:
                        window.after(delay_sec, send)
                    else:
                        smtp_engine.complete()
                        progress_label.config(
                            text=f"Completed sending to {smtp_engine.count} recipients"
                        )
                else:
                    smtp_engine.msg_type_err()
            # Catch possible incoming errors
# SmtpEngine holds mails, count, sent
# Default subject and sender used when empty
# Relay delay helps avoid rate limiting
# Credentials validated before send
# Message type: Html or Plaintext
# Default subject and sender used when empty
# Mail list can be changed via Load Mail List button
# Test Connection verifies SMTP before sending
# ScrolledText used for long messages
# Port 465 uses SSL; 587/25 use STARTTLS
            except IndexError:
                smtp_engine.mail_error()
            except smtplib.SMTPAuthenticationError:
                smtp_engine.connection_error()
            except smtplib.SMTPSenderRefused:
                smtp_engine.refuse_err()
            except smtplib.SMTPConnectError:
                smtp_engine.smtp_error()
    else:
        smtp_engine.option_invalid()


def test_connection():
    """Test SMTP connection without sending a message."""
    validate = "Select an Option"
    user = username_entry.get()
    passwd = password_entry.get()
    server = outgoing_server.get()
    port = port_type.get()

    if port == validate:
        smtp_engine.option_invalid()
        return

    if len(user) == 0 or len(passwd) == 0 or len(server) == 0:
        smtp_engine.no_field()
        return

    smtp_engine.test_connection(user, passwd, server, port)


# Creating the buttons
# Port 465 uses SSL; 587/25 use STARTTLS
# Mail list can be changed via Load Mail List button
# Relay delay in milliseconds for window.after
clear_message = Button(
    text="Send Message",
    highlightbackground=WHITE,
    fg=BLACK,
    width=17,
    command=send,
)
clear_message.grid(column=3, row=12)

test_button = Button(
    text="Test Connection",
    highlightbackground=WHITE,
    fg=BLACK,
    width=17,
    command=test_connection,
)
test_button.grid(column=2, row=12)

load_list_button = Button(
    text="Load Mail List",
    highlightbackground=WHITE,
    fg=BLACK,
    width=17,
    command=choose_mail_list,
)
load_list_button.grid(column=0, row=7, pady=5)

# Load the mail image
# Credentials validated before send
canvas = Canvas(width=200, height=200, bg=WHITE, highlightthickness=0)
photo = PhotoImage(file="logo.png")
canvas.create_image(100, 100, image=photo)
canvas.grid(column=1, row=0, columnspan=2)
window.mainloop()
